pkg/module/fetch.go

```diff
@@ -0,0 +1,32 @@
+package module
+
+import (
+	"context"
+
+	"github.com/gomods/athens/pkg/storage"
+)
+
+// Fetch downloads the module@version using the fetcher and stores it in storage
+func Fetch(ctx context.Context, s storage.Backend, fetcher Fetcher, mod, version string, mf *Filter) error {
+	if !mf.ShouldProcess(mod) {
+		return NewErrModuleExcluded(mod)
+	}
+
+	if s.Exists(ctx, mod, version) {
+		return NewErrModuleAlreadyFetched(mod, version)
+	}
+
+	moduleRef, err := fetcher.Fetch(mod, version)
+	if err != nil {
+		return err
+	}
+	defer moduleRef.Clear()
+
+	// pretend like moduleLoc has $version.info, $version.mod and $version.zip in it :)
+	module, err := moduleRef.Read()
+	if err != nil {
+		return err
+	}
+
```

Comment:
Shouldn't we do defer module.Zip.Close() here? Save doesn't call Close AFAIK.

Reply:
What will happen if Clear() is called before zip gets closed? What is the defer calls order?

Reply:
Good catch - yes, we need to call it. `defer`s execute in stack order, so the zip would be closed before the clear. I wrote an example to verify: https://play.golang.org/p/Q0x563MPJoi